src/SAWScript/Crucible/LLVM/MethodSpecIR.hs

```diff
@@ -597,3 +610,77 @@ mkSomeLLVM x = Some (Compose x)
 
 getSomeLLVM :: forall t. (forall arch. t (LLVM arch)) -> AllLLVM t
 getSomeLLVM x = All (Compose x)
+
+--------------------------------------------------------------------------------
+-- *** ResolvedState
+
+type instance MS.ResolvedState (LLVM arch) = LLVMResolvedState
+
+data ResolvedPathItem
+  = ResolvedField String
+  | ResolvedElem Int
+  | ResolvedCast L.Type
+ deriving (Show, Eq, Ord)
+
+type ResolvedPath = [ResolvedPathItem]
+
+-- | A datatype to keep track of which parts of the simulator state -- have been initialized already. For each allocation unit or global,
+-- we keep a list of element-paths that identify the initialized
+-- sub-components.
+data LLVMResolvedState =
+  ResolvedState
+    { _rsAllocs :: Map MS.AllocIndex [ResolvedPath]
+    , _rsGlobals :: Map String [ResolvedPath]
+    }
+  deriving (Eq, Ord, Show)
+
+emptyResolvedState :: LLVMResolvedState
+emptyResolvedState = ResolvedState Map.empty Map.empty
+
+makeLenses ''LLVMResolvedState
+
+-- | Record the initialization of the pointer represented by the given
+-- SetupValue.
+markResolved ::
+  MS.SetupValue (LLVM arch) ->
+  ResolvedPath {-^ path within this object (if any) -} ->
+  LLVMResolvedState ->
+  LLVMResolvedState
+markResolved val0 path0 rs = go path0 val0
+  where
+    go path val =
+      case val of
+        MS.SetupVar n         -> rs & rsAllocs %~ Map.alter (ins path) n
+        MS.SetupGlobal _ name -> rs & rsGlobals %~ Map.alter (ins path) name
+        MS.SetupElem _ v idx  -> go (ResolvedElem idx : path) v
+        MS.SetupField _ v fld -> go (ResolvedField fld : path) v
+        MS.SetupCast _ v tp   -> go (ResolvedCast tp : path) v
+        _                  -> rs
+
+    ins path Nothing = Just [path]
+    ins path (Just paths) = Just (path : paths)
+
+-- | Test whether the pointer represented by the given SetupValue has
+-- been initialized already.
+testResolved ::
+  MS.SetupValue (LLVM arch) ->
+  ResolvedPath {-^ path within this object (if any) -} ->
+  LLVMResolvedState ->
+  Bool
+testResolved val0 path0 rs = go path0 val0
+  where
+    go path val =
+      case val of
+        MS.SetupVar n         -> test path (Map.lookup n (_rsAllocs rs))
+        MS.SetupGlobal _ c    -> test path (Map.lookup c (_rsGlobals rs))
+        MS.SetupElem _ v idx  -> go (ResolvedElem idx : path) v
+        MS.SetupField _ v fld -> go (ResolvedField fld : path) v
+        MS.SetupCast _ v tp   -> go (ResolvedCast tp : path) v
+        _                  -> False
+
+    test _ Nothing = False
+    test path (Just paths) = any (overlap path) paths
+
+    overlap (x : xs) (y : ys) = x == y && overlap xs ys
```

Comment:
If I'm reading this correctly, `LLVMResolvedState`s that contain `ResolvedCast`s will only be considered equal if the `Type`s they contain are syntactically equal? If so, this seems like it could be prone to false negatives if the user isn't careful about things like `typedef`s. Then again, I don't think this function was ever 100% foolproof—perhaps worth documenting that.

Reply:
Indeed, it's my understanding that this test is a best-effort check for the purposes of providing user-actionable feedback.